cmd/silkworm.cpp

```diff
@@ -14,291 +14,109 @@
-                    "If item is NOT in the list - means NO pruning for this data.\n"
-                    "Example: --prune=hrtc (default: none)",
-                    true)
-        ->check(PruneModeValidator());
+#include "common.hpp"
 
-    prune_opts.add_option("--prune.h.older", "Override default 90k blocks of history to prune")
-        ->check(CLI::Range(0u, UINT32_MAX));
-    prune_opts.add_option("--prune.r.older", "Override default 90k blocks of receipts to prune")
-        ->check(CLI::Range(0u, UINT32_MAX));
-    prune_opts.add_option("--prune.t.older", "Override default 90k blocks of transactions to prune")
-        ->check(CLI::Range(0u, UINT32_MAX));
-    prune_opts.add_option("--prune.c.older", "Override default 90k blocks of call traces to prune")
-        ->check(CLI::Range(0u, UINT32_MAX));
-    prune_opts.add_option("--prune.h.before", "Prune history data before this block")
-        ->check(CLI::Range(0u, UINT32_MAX));
-    prune_opts.add_option("--prune.r.before", "Prune receipts data before this block")
-        ->check(CLI::Range(0u, UINT32_MAX));
-    prune_opts.add_option("--prune.t.before", "Prune transactions data before this block")
-        ->check(CLI::Range(0u, UINT32_MAX));
-    prune_opts.add_option("--prune.c.before", "Prune call traces data before this block")
-        ->check(CLI::Range(0u, UINT32_MAX));
-
-    // Logging options
-    auto& log_opts = *cli.add_option_group("Log", "Logging options");
-    log_opts.add_option("--log.verbosity", log_settings.log_verbosity, "Sets log verbosity", true)
-        ->check(CLI::Range(static_cast<uint32_t>(log::Level::kCritical), static_cast<uint32_t>(log::Level::kTrace)))
-        ->default_val(std::to_string(static_cast<uint32_t>(log_settings.log_verbosity)));
-    log_opts.add_flag("--log.stdout", log_settings.log_std_out, "Outputs to std::out instead of std::err");
-    log_opts.add_flag("--log.nocolor", log_settings.log_nocolor, "Disable colors on log lines");
-    log_opts.add_flag("--log.utc", log_settings.log_utc, "Prints log timings in UTC");
-    log_opts.add_flag("--log.threads", log_settings.log_threads, "Prints thread ids");
-    log_opts.add_option("--log.file", log_settings.log_file, "Tee all log lines to given file name");
+using namespace silkworm;
 
-    cli.parse(argc, argv);
+static std::atomic<size_t> s_allocated_memory{0};
 
-    // Assign settings
-    node_settings.data_directory = std::make_unique<DataDirectory>(datadir, /*create=*/true);
-    node_settings.chaindata_env_config.max_size = parse_size(chaindata_max_size).value();
-    node_settings.chaindata_env_config.growth_size = parse_size(chaindata_growth_size).value();
-    if (node_settings.chaindata_env_config.growth_size > node_settings.chaindata_env_config.max_size / 2) {
-        throw std::invalid_argument("--chaindata.growthsize too wide");
-    }
+void* operator new(size_t size) {
+    s_allocated_memory += size;
+    return malloc(size);
+}
```

Comment:
I agree with what @yperbasis  wrote. In addition The C++ standard specifies that operator new [throws std::bad_alloc or another exception derived from std::bad_alloc (since C++11) on failure to allocate memory](https://en.cppreference.com/w/cpp/memory/new/operator_new), in addition to other requirements such as retries and alignment. This overload doesn't implement the specification.

Reply:
This is correct in principle but not in real life.
Modern OSes allow over-allocation and both `new` and `malloc` can return valid pointers even if there's not enough memory left hence `std::bad_alloc` might not throw and cannot be considered as a reliable sign of memory exhaustion. What **it does happen instead** is that if you try to **access** memory pointed by an over-allocated segment you get a segfault which is not catcheable (you can handle the signal but cannot resume the program).

Reply:
Removed by 1fef176